### PR TITLE
Behave like cudart wrt. threads.

### PR DESCRIPTION
**Pull request by maleadt**

### CODE REVIEW

File: src/init.jl [hunk 1]
```diff
@@ -71,7 +71,9 @@ function device!(dev::CuDevice)
         CuContext(pctx)
         pop!(CuContext) # we don't maintain a stack...
     end
-    activate(ctx)       # replace the top of the stack
+    Threads.@threads for i in 1:Threads.nthreads()
```
Comment (vchuravy) [inline src/init.jl]: Hm, this adds quite a bit of overhead to something we want to be a low-cost operation. 

Additionally I could see a case where users want to do:

```
@threads for i in 1:nthreads()
   device!(CuDevice(i))
end
```

But that my be a future us problem. Is right now the design criteria that cuda calls from task should just work and
they all end up on the same device?

Comment (maleadt) [reply src/init.jl]: Hmm right that's a valid concern. We should probably extend the lazy-initialization to a per-thread initialization then.

>  Is right now the design criteria that cuda calls from task should just work and
> they all end up on the same device?

Basically yes, but you're right that it should be possible to bind contexts to separate threads. That should be possible with the extended initialization, and not with the current design. I'll have another go.

**Comment by maleadt** [comment]
https://devblogs.nvidia.com/cuda-pro-tip-always-set-current-device-avoid-multithreading-bugs/
So we should pick the default device again for each thread.

**Comment by maleadt** [comment]
Demo

```
$ JULIA_DEBUG=CUDAnative JULIA_NUM_THREADS=2 julia   
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.3.0 (2019-11-26)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> using CUDAnative, CUDAdrv


# function to report on the state of threads

julia> report() = Threads.@threads for i in 1:Threads.nthreads()
         println("Thread $(Threads.threadid()): ", CuCurrentContext(), CuCurrentContext()===nothing ? "" : " for device $(device())")
       end

julia> report()
Thread 1: nothing
Thread 2: nothing


# let's lazy-initialize the first thread

julia> CUDAdrv.total_memory()
┌ Debug: Initializing CUDA on thread 1 after call to cuMemGetInfo
└ @ CUDAnative ~/Julia/pkg/CUDAnative/src/init.jl:47
11523260416

julia> report()
Thread 2: nothing
Thread 1: CuContext(Ptr{Nothing} @0x0000000002331c60) for device CuDevice(0): GeForce RTX 2080 Ti


# switching devices only affects a single thread

julia> device!(1)

julia> report()
Thread 2: nothing
Thread 1: CuContext(Ptr{Nothing} @0x0000000002437610) for device CuDevice(1): GeForce GTX 1080


# let's initialize all threads

julia> Threads.@threads for i in 1:Threads.nthreads()
         println(CUDAdrv.total_memory())
       end
┌ Debug: Initializing CUDA on thread 2 after call to cuMemGetInfo
└ @ CUDAnative ~/Julia/pkg/CUDAnative/src/init.jl:47
11523260416
8512602112

julia> report()
Thread 2: CuContext(Ptr{Nothing} @0x0000000002331c60) for device CuDevice(0): GeForce RTX 2080 Ti
Thread 1: CuContext(Ptr{Nothing} @0x0000000002437610) for device CuDevice(1): GeForce GTX 1080


# just to show that switching only really affects a single thread

julia> device!(0)

julia> report()
Thread 1: CuContext(Ptr{Nothing} @0x0000000002331c60) for device CuDevice(0): GeForce RTX 2080 Ti
Thread 2: CuContext(Ptr{Nothing} @0x0000000002331c60) for device CuDevice(0): GeForce RTX 2080 Ti

julia> device!(1)

julia> report()
Thread 2: CuContext(Ptr{Nothing} @0x0000000002331c60) for device CuDevice(0): GeForce RTX 2080 Ti
Thread 1: CuContext(Ptr{Nothing} @0x0000000002437610) for device CuDevice(1): GeForce GTX 1080


# overhead

julia> using BenchmarkTools

julia> @btime CuCurrentContext()
  108.284 ns (0 allocations: 0 bytes)
CuContext(Ptr{Nothing} @0x0000000002437610)

# vs
julia> handle_ref = Ref{CUDAdrv.CUcontext}()
Base.RefValue{Ptr{Nothing}}(Ptr{Nothing} @0x4076cee38e38e38e)
julia> @btime CUDAdrv.cuCtxGetCurrent($handle_ref)
  66.820 ns (0 allocations: 0 bytes)

julia> @btime device!(0)
  349.917 ns (1 allocation: 32 bytes)
```

@vchuravy acceptable performance/overhead?

**Comment by maleadt** [comment]
With a fast path and using CuDevice directly for benchmarking (Int->CuDevice API is apparently slow):
```
$ julia -e 'using BenchmarkTools, CUDAnative, CUDAdrv; display(@btime device!($(CuDevice(0))))'
  50.912 ns (0 allocations: 0 bytes)
```

**Comment by maleadt** [comment]
Segfault on nightly@cyclops:

```
signal (11): Segmentation fault
in expression starting at none:7
unknown function (ip: 0x7f0fbe248fef)
unknown function (ip: 0x7f0fbe3f40a2)
unknown function (ip: 0x7f0fbe3f5fb8)
unknown function (ip: 0x7f0fbe3f6a2d)
unknown function (ip: 0x7f0fbe3f6df2)
unknown function (ip: 0x7f0fbe24b95a)
unknown function (ip: 0x7f0fbe22042e)
unknown function (ip: 0x7f0fbe231480)
unknown function (ip: 0x7f0fbe37b390)
unknown function (ip: 0x7f0fbe37b446)
unknown function (ip: 0x7f0fbe10c093)
unknown function (ip: 0x7f0fbe113bba)
unknown function (ip: 0x7f0fbe117a0d)
__cuda_CallJitEntryPoint at /usr/lib/x86_64-linux-gnu/libnvidia-ptxjitcompiler.so.1 (unknown line)
nvPTXCompilerCompile at /usr/lib/x86_64-linux-gnu/libnvidia-ptxjitcompiler.so.1 (unknown line)
fatBinaryCtl_Compile at /usr/lib/x86_64-linux-gnu/libnvidia-fatbinaryloader.so.418.56 (unknown line)
unknown function (ip: 0x7f0fdb3346f3)
unknown function (ip: 0x7f0fdb335211)
unknown function (ip: 0x7f0fdb2489b4)
cuModuleLoadDataEx at /usr/lib/x86_64-linux-gnu/libcuda.so (unknown line)
macro expansion at /builds/JuliaGPU/CUDAnative.jl/.julia/packages/CUDAapi/CCgJL/src/call.jl:40 [inlined]
macro expansion at /builds/JuliaGPU/CUDAnative.jl/.julia/packages/CUDAdrv/iEACC/src/error.jl:118 [inlined]
cuModuleLoadDataEx at /builds/JuliaGPU/CUDAnative.jl/.julia/packages/CUDAdrv/iEACC/src/libcuda.jl:232
macro expansion at ./gcutils.jl:105 [inlined]
CuModule at /builds/JuliaGPU/CUDAnative.jl/.julia/packages/CUDAdrv/iEACC/src/module.jl:34
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2158 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2322
macro expansion at /builds/JuliaGPU/CUDAnative.jl/src/execution.jl:424 [inlined]
#cufunction#198 at /builds/JuliaGPU/CUDAnative.jl/src/execution.jl:359
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2158 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2322
cufunction at /builds/JuliaGPU/CUDAnative.jl/src/execution.jl:359
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2158 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2322
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1692 [inlined]
do_call at /buildworker/worker/package_linux64/build/src/interpreter.c:369
eval_value at /buildworker/worker/package_linux64/build/src/interpreter.c:458
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:682
jl_interpret_toplevel_thunk at /buildworker/worker/package_linux64/build/src/interpreter.c:913
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:814
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:764
jl_toplevel_eval_in at /buildworker/worker/package_linux64/build/src/toplevel.c:843
eval at ./boot.jl:331
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2144 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2322
exec_options at ./client.jl:264
_start at ./client.jl:484
jfptr__start_2076.clone_1 at /usr/local/lib/julia/sys.so (unknown line)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2144 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2322
jl_apply at /buildworker/worker/package_linux64/build/ui/../src/julia.h:1692 [inlined]
true_main at /buildworker/worker/package_linux64/build/ui/repl.c:96
main at /buildworker/worker/package_linux64/build/ui/repl.c:217
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
_start at /usr/local/bin/julia (unknown line)
Allocations: 30877995 (Pool: 30871360; Big: 6635); GC: 35
```

**Comment by vchuravy** [comment]
https://github.com/JuliaGPU/CUDAnative.jl/pull/514#issuecomment-567014045 looks good :)

**Comment by maleadt** [comment]
We can do better

```
julia> using BenchmarkTools

julia> @btime CuCurrentContext()
  45.930 ns (0 allocations: 0 bytes)
CuContext(Ptr{Nothing} @0x0000000001846510)

julia> @btime device!(0)
  53.901 ns (0 allocations: 0 bytes)
```

**Comment by maleadt** [comment]
```
julia> @btime device!($(CuDevice(1))) setup=(device!(CuDevice(0)))
  48.802 ns (0 allocations: 0 bytes)
```

Solid improvement, but I'm not sure if it's enough to prefix every operation with. We need to think about how to go forwards with multithreading: have any operation seamlessly work (in which case they basically would need to switch contexts if necessary), or whether we want to impose this on the user to switch contexts at the beginning of a blob of code that might run on a different thread (or might use a different device -- there's a lot of parallels to be drawn between multithreading and multigpu).

**Pull request merged**